docs-source/content/userguide/cicd/_index.md

```diff
@@ -0,0 +1,65 @@
+---
+title: "CI/CD"
+date: 2019-04-11T13:01:55-04:00
+weight: 5
+description: "Learn about managing domain images with CI/CD."
+draft: false
+---
+
+### Overview
+
+In this section, we will discuss the recommended techniques for managing the evolution
+and mutation of Docker images to run WebLogic in Kubernetes.  There are several
+approaches and techniques available, and the choice of which to use depends very
+much on your particular requirements.  We will start with a review of the "problem
+space," and then talk about the considerations that would lead us to choose various
+approaches.  We will provide details about several approaches to implementing
+CI/CD and links to examples.
+
+### Review of the problem space
+
+Kubernetes makes a fundamental assumption that Docker images are immutable,
+that they contain no state, and that updating them is as simple as throwing
+away a pod/container and replacing it with a new one that uses a newer version
+of the Docker image.  These assumptions work very well for microservices
+applications, but for more traditional workloads, we need to do some extra
+thinking and some extra work to get the behavior we want.
+
+CI/CD is an area where the standard assumptions aren't always suitable.  In the
+microservices architecture, you typically minimize dependencies and build
+images from scratch with all of the dependencies in them.  You also typically
+keep all of the configuration outside of the image, for example, in Kubernetes config
+maps or secrets, and all of the state outside of the image too.  This makes
+it very easy to update running pods with a new image.
+
+Let's consider how a WebLogic image is different.  There will, of course, be a
+base layer with the operating system; let's assume it is
+[Oracle Linux "slim"](https://hub.docker.com/_/oraclelinux/).  Then you need
+a JDK and this is very commonly in another layer.  Many people will use
+the officially supported JDK images from the Docker Store, like the
+[Server JRE image](https://hub.docker.com/_/oracle-serverjre-8), for example.  On
+top of this, you need the WebLogic Server binaries (the "Oracle Home").  On top
+of that, you may wish to have some patches or updates installed.  And then
+you need your domain, that is the configuration.
+
+There is also other information associated with a domain that needs to live
+somewhere, for example leasing tables, message and transaction stores, and so
+on.  We recommend that these be kept in a database to take advantage of built-in 
+database server HA, and the fact that disaster recovery of sites across all 
+but the shortest distances almost always requires using a single database 
+server to consolidate and replicated data (DataGuard).
+
+There are two common approaches on how to structure these components.  The first,
```

Comment:
It just occur to me that is it might be helpful to introduce a third approach: a hybrid of domaim-on-a-pv and domain-in-image. 
1. Create and update a domain to a pv, probably via `docker run`.
1. When come to build the domain image, just copy the domainHome from a PV to the image.

In this way we can keep all the benefits of domain-in-image while get rid of regenerating the domain when rebuild the domain image.
With this approach we can mutate the domain layer while keep the same domain encryption keys.

Reply:
good idea @lilyhe123 - @rjeberhard  and I were actually discussing the same thing last week, and I think @rjeberhard  is working on testing it.  maybe you could help to test it?

Reply:
I am not sure customers will be happy with this option since they will need to maintain both domain image and PV.   The second problem with this approach is that it goes against Docker best practices.   I would prefer to spend the effort in enhancing WebLogic to export the key so that then a new domain image can be create using the same key.

Reply:
Not clear what implications this would have since you are taking the domain and trying to copy that domain at various snapshots, are the server's booting while on PV?

Reply:
@markxnelson I'm sure basic thing should work with these hybrid approach but we need to brainstorm the downsides of it. Monica already have one.

@mriccell I agree that this is only a workaround. If we can enhance Weblogic to regenerate a domain without a key. That will be great!  But what Docker best practice you mean here?

@TheFrogPad the copy is done when creating the image, not in the running container.

Reply:
@lilyhe123 maybe I misunderstood "copy the domainHome from a PV to the image" do you mean build the image by doing a copy of the domain from the PV in the Dockerfile?

Reply:
I still not thrilled with this 3rd option just the fact of having to create both a domain inside of the image and then have to create the PV and persist the domain to a PV.  I think just archiving the domain, applications and then creating a new image from the archived domain is less steps and simpler for the user until we have the WebLogic enhancement.

Reply:
@lilyhe123 @mriccell What is meant "regenerate domain without a key"?

Two WLS domains are not the same from an administrative and security perspective, I do no think anyone would expect that the administrator of another domain could simply come to your domain and make changes without valid credentials in your domain and one would certainly not expect that a server from another domain would join your domain unexpectedly as well?

Perhaps like Monica I'm unclear of the benefit of a PV vs creating an archive with the primordial domain that is re-used as application updates are made?

Reply:
Correction: "regenerate a domain without changing the encryption key".

Reply:
wait/defer

Reply:
If you talk about some manual steps to avoid regenerating the domain and keep the same keys, then any approach that can achieve it is fine, e.g. archive and recreate. 

Just one thing I want to add. Because of the nature of docker image layers, if the lower image is updated all the upper images that depend on it need to be rebuilt. Ideally the CI/CD system needs to find a way to do this **automatically**. That means the same Dockerfile need to be run when the image is first created and any time need to do rebuilding.